### PR TITLE
HELIUM-414: Port 0 re-use

### PR DESCRIPTION
**Pull request by macpie**

### CODE REVIEW

File: test/listen_SUITE.erl [hunk 1]
```diff
@@ -61,10 +105,88 @@ already_test(Config) ->
...
+%% @end
+%%--------------------------------------------------------------------
+port0_reuse(_Config) ->
+    SwarmOpts = [{libp2p_transport_tcp, [{nat, false}]}],
+
+    {ok, Swarm} = libp2p_swarm:start(listen_port0_reuse, SwarmOpts),
+    ok = libp2p_swarm:listen(Swarm, "/ip4/0.0.0.0/tcp/0"),
+    ok = libp2p_swarm:listen(Swarm, "/ip6/::/tcp/0"),
+
+    timer:sleep(2000),
```
Comment (madninja) [inline test/listen_SUITE.erl]: Why do we need to sleep here?

File: test/listen_SUITE.erl [hunk 2]
```diff
@@ -61,10 +105,88 @@ already_test(Config) ->
...
+    ListenAddrs = libp2p_swarm:listen_addrs(Swarm),
+
+    ok = libp2p_swarm:stop(Swarm),
+    timer:sleep(2000),
```
Comment (madninja) [inline test/listen_SUITE.erl]: Ditto? Is this assuming how long it takes to tear down the swarm? It seems very brittle?

File: test/listen_SUITE.erl [hunk 3]
```diff
@@ -61,10 +105,88 @@ already_test(Config) ->
...
+    {ok, Swarm4} = libp2p_swarm:start(listen_port0_reuse2, SwarmOpts),
+    ok = libp2p_swarm:listen(Swarm4, "/ip4/127.0.0.1/tcp/0"),
+    ok = libp2p_swarm:listen(Swarm4, "/ip6/::1/tcp/0"),
+
+    timer:sleep(2000),
+    ListenAddrs2 = libp2p_swarm:listen_addrs(Swarm4),
+    ok = libp2p_swarm:stop(Swarm4),
+    ok.
+
+%% ------------------------------------------------------------------
```
Comment (madninja) [inline test/listen_SUITE.erl]: Delete the below?

File: src/libp2p_transport_tcp.erl [hunk 4]
```diff
@@ -343,9 +343,14 @@ listen_on(Addr, TID) ->
                                               DefaultListenOpts),
             % Dialyzer severely dislikes ranch_tcp:listen so we
             % emulate it's behavior here
+            Port = reuseport0(TID, Type, IP, Port0),
             case gen_tcp:listen(Port, [Type | AddrOpts] ++ ListenOpts) of
                 {ok, Socket} ->
                     ListenAddrs = tcp_listen_addrs(Socket),
+
+                    Cache = libp2p_swarm_sup:cache(TID),
```
Comment (Vagabond) [inline src/libp2p_transport_tcp.erl]: Aren't most of these types of calls wrapped by a function in libp2p_swarm? eg `libp2p_swarm:cache()`

Comment (macpie) [reply src/libp2p_transport_tcp.erl]: I did like peer book `libp2p_swarm_sup:register_peerbook` and `libp2p_swarm_sup:peerbook`

Comment (madninja) [reply src/libp2p_transport_tcp.erl]: Right but there's _also_ a libp2p_swarm:peerbook function that calls the _sup function

The idea is that libp2p_swarm is the supported API of the swarm. The sup is just an implementation tactic.

Could you add that cache accessor in libp2p_swarm (including an @doc if you could) @macpie ?

**Pull request merged**